### PR TITLE
[BEAM-7060] Type hints from Python 3 annotations

### PR DESCRIPTION
**Pull request by udim**
Implements IOTypeHints.from_callable() to get type hints where
annotations are available.

Some under the hood changes to how arguments are obtained and applied. See commits for more details.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
Hi @robertwb, I've been working on this for a while and I'd like to get any input you have on it. Still has plently of TODOs and probably some bugs (some TODOs might be out of date).

File: sdks/python/apache_beam/typehints/typehints_test.py [hunk 1]
```diff
@@ -918,139 +776,16 @@ def test_must_be_single_return_type(self):
...
-  def test_type_check_simple_type(self):
-    @with_output_types(str)
-    def upper(a):
-      return a.upper()
-    self.assertEqual('TEST', upper('test'))
-
   def test_type_check_composite_type(self):
     @with_output_types(typehints.List[typehints.Tuple[int, int]])
-    def bar():
+    def unused_bar():
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typehints_test.py]: What exactly is this testing?

File: sdks/python/apache_beam/typehints/typehints_test.py [hunk 2]
```diff
@@ -918,139 +776,16 @@ def test_must_be_single_return_type(self):
...
   def test_any_return_type_hint(self):
     @with_output_types(typehints.Any)
-    def bar():
+    def unused_bar():
       return 'foo'
 
-    self.assertEqual('foo', bar())
-
-
-class CombinedReturnsAndTakesTestCase(TypeHintTestCase):
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typehints_test.py]: Could you explain why these were all deleted?

File: sdks/python/apache_beam/typehints/typehints_test.py [hunk 3]
```diff
@@ -1062,57 +797,92 @@ def test_hint_helper(self):
...
-        getcallargs_forhints(func, *[Any, Any]))
+        getcallargs_forhints(False, func, *[Any, Any]))
+    if sys.version_info >= (3,):
+      self.assertEqual(
+          {'a': Any, 'b_c': Any, 'd': Tuple[Union[int, str], ...]},
+          getcallargs_forhints(False, func, *[Any, Any, str, int]))
+    else:
+      self.assertEqual(
+          {'a': Any, 'b_c': Any, 'd': Tuple[Any, ...]},
+          getcallargs_forhints(False, func, *[Any, Any, Any, int]))
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typehints_test.py]: Why are we testing something different here (and below)?

File: sdks/python/apache_beam/typehints/decorators.py [hunk 4]
```diff
@@ -265,23 +368,34 @@ def _unpack_positional_arg_hints(arg, hint):
...
 
-def getcallargs_forhints(func, *typeargs, **typekwargs):
-  """Like inspect.getcallargs, but understands that Tuple[] and an Any unpack.
+def getcallargs_forhints(using_var_hints, func, *typeargs, **typekwargs):
+  """Like inspect.getcallargs, with support for declaring default args as Any.
+
+  In Python 2, understands that Tuple[] and an Any unpack.
+
+  Args:
+    using_var_hints: For variable length arguments, whether to expect the bound
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/decorators.py]: I have to admit, I find this boolean setting quite hard to follow, even looking where it's used (once True and once False). What is the overall goal we're trying to achieve with this flexibility?

File: sdks/python/apache_beam/typehints/typehints_test_py3.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import unittest
+
+from apache_beam.transforms.core import DoFn
+from apache_beam.typehints import KV
+
+
+class TestParDoAnnotations(unittest.TestCase):
+  def test_with_side_input(self):
+    class MyDoFn(DoFn):
+      def process(self, element: float, side_input: str) -> KV[str, float]:
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typehints_test_py3.py]: This will be so nice :)!

File: sdks/python/apache_beam/typehints/decorators.py (hunk 4 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/decorators.py]: getcallargs_forhints is being passed two variations of values for variadic arguments and it needs to be able to tell them apart.
After calling `signature.bind()`, getcallargs_forhints sees a tuple or a dict (for VAR_POSITIONAL or VAR_KEYWORD), which might be wrapping a well formed hint for that argument or one or more hints that need to be combined into a Tuple or Dict.

For example, it might see a `VAR_POSITIONAL` with a value of `tuple(int, str)` and need to convert that to `Tuple[Union[int, str], ...]`, or it might see a value of `tuple(Tuple[Any, ...])` and just need to unwrap it to `Tuple[Any, ...]`.

File: sdks/python/apache_beam/typehints/typehints_test.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/typehints_test.py]: I removed the tests that used `@check_type_hints`, since that decorator doesn't exercise the actual type checking code paths. `PTransformTypeCheckTestCase` has coverage for `runtime_type_check = True`.

File: sdks/python/apache_beam/typehints/typehints_test.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/typehints_test.py]: Not sure. Maybe the original author intended to add the `@check_type_hints` decorator to the function? (same for `test_any_return_type_hint` below)

Comment (udim) [reply sdks/python/apache_beam/typehints/typehints_test.py]: Or more likely it's verifying that a composite type is accepted as a type hint.

File: sdks/python/apache_beam/typehints/typehints_test.py (hunk 3 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/typehints_test.py]: The Python 2 version (`getcallargs_forhints_impl_py2`) has a bug in it that always sets varargs to `Tuple[Any, ...]`. I left in a TODO since I didn't want to complicate this PR any further.

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
python 3.7 postcommit

**Comment by udim** [comment]
@robertwb PTAL

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
Run Portable_Python PreCommit

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
Run Python 2 PostCommit

**Comment by udim** [comment]
run python 3.5 postcommit

**Comment by udim** [comment]
run python 3.6 postcommit

**Comment by udim** [comment]
run python 2.7 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
run python 2 postcommit

File: sdks/python/apache_beam/typehints/decorators.py (hunk 4 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/decorators.py]: I think we might be able to auto-detect the variation in use.

For VAR_POSITIONAL:
- True if we got tuple(Tuple[FOO, ...]), otherwise False

For VAR_KEYWORD:
- True if we got tuple(Dict[FOO, BAR]), otherwise False

**Comment by udim** [comment]
run python 2.7 postcommit

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by udim** [comment]
run python 2 postcommit

File: sdks/python/apache_beam/transforms/core.py [hunk 6]
```diff
@@ -638,7 +665,13 @@ def __repr__(self):
     return 'CallableWrapperDoFn(%s)' % self._fn
 
   def default_type_hints(self):
-    type_hints = get_type_hints(self._fn)
+    fn_type_hints = typehints.decorators.IOTypeHints.from_callable(self._fn)
+    if fn_type_hints is not None:
+      try:
+        fn_type_hints.strip_iterable()
+      except ValueError as e:
+        raise ValueError('Return value not iterable: %s: %s' % (self, e))
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: (self._fn, e)

File: sdks/python/apache_beam/typehints/decorators.py [hunk 7]
```diff
@@ -171,22 +269,55 @@ def simple_output_type(self, context):
...
+    else:
+      input_types = hints.input_types
+    if self._has_output_types():
+      output_types = self.output_types
+    else:
+      output_types = hints.output_types
+    return IOTypeHints(input_types, output_types)
+
+  def _has_input_types(self):
+    return self.input_types is not None and any(self.input_types)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/typehints/decorators.py]: Ah. Probably too much to clean up in this CL, but it seems it'd be good to normalize `((), {})` to None everywhere (or the other way around) or possibly even use a class to represent this that a raw tuple.

**Comment by udim** [comment]
The `test_streaming_with_attributes` failures are unrelated to this PR. I opened https://issues.apache.org/jira/browse/BEAM-8153 for those

File: sdks/python/apache_beam/transforms/core.py (hunk 6 above)
Comment (udim) [reply sdks/python/apache_beam/transforms/core.py]: goot catch!

File: sdks/python/apache_beam/typehints/decorators.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/typehints/decorators.py]: Will do. I don't recall anymore why I wanted this, but I do remember that I would need to touch a lot of places that check for None. Opened https://issues.apache.org/jira/browse/BEAM-8154 to track

**Comment by udim** [comment]
run python 3.5 postcommit

**Comment by udim** [comment]
run python 2 postcommit

**Comment by udim** [comment]
Run Python 2 PostCommit
